src/main/java/com/flansmod/common/guns/EntityBullet.java

```diff
@@ -681,24 +694,54 @@ else if (motionZ != 0)
                         worldObj.spawnEntityInWorld(new EntityDebugDot(worldObj, new Vector3f(posX + motionX * entityHit.intersectTime, posY + motionY * entityHit.intersectTime, posZ + motionZ * entityHit.intersectTime), 1000, 1F, 1F, 0F));
                         FlansMod.log(entityHit.entity.toString() + ": d=" + d + ": damage=" + damage + ": type.damageVsEntity=" + type.damageVsEntity);
                     }    
-                } else if (bulletHit instanceof BlockHit) {
+                } else if (bulletHit instanceof BlockHit) {             	
                     BlockHit blockHit = (BlockHit) bulletHit;
-                    MovingObjectPosition raytraceResult = blockHit.raytraceResult;
+                    MovingObjectPosition raytraceResult = blockHit.raytraceResult;                    
+                    Vec3 hitVec = raytraceResult.hitVec;
+                    
                     //If the hit wasn't an entity hit, then it must've been a block hit
                     int xTile = raytraceResult.blockX;
                     int yTile = raytraceResult.blockY;
                     int zTile = raytraceResult.blockZ;
+                    
+                    
                     if (FlansMod.DEBUG)
-                        worldObj.spawnEntityInWorld(new EntityDebugDot(worldObj, new Vector3f(raytraceResult.hitVec.xCoord, raytraceResult.hitVec.yCoord, raytraceResult.hitVec.zCoord), 1000, 0F, 1F, 0F));
+                        worldObj.spawnEntityInWorld(new EntityDebugDot(worldObj, new Vector3f(hitVec.xCoord, hitVec.yCoord, hitVec.zCoord), 1000, 0F, 1F, 0F));
                     
                     Block block = worldObj.getBlock(xTile, yTile, zTile);
-                    Material mat = block.getMaterial();
+                    Material mat = block.getMaterial();             
+                    
+                    if(FlansMod.enableBlockPenetration) {
+                    	for(Block penetrableBlock : FlansMod.penetrableBlocks.keySet()) { 
+                        	if(block != penetrableBlock) continue;
+                        	
+                        	Float[] valuesArray = FlansMod.penetrableBlocks.get(penetrableBlock);
+                        	float metadata = valuesArray[0];
+                        	
+                        	if(metadata != -1 && metadata != worldObj.getBlockMetadata(xTile, yTile, zTile)) continue;
+                        	
+                        	float hardness = valuesArray[1];
+                        	
+                        	penetratingPower-=hardness * (type.blockPenetrationModifier > 0 ? (1F/type.blockPenetrationModifier) : 1F);
+                        	if(penetratingPower < 0) break;
+                        	
+                        	FlansMod.proxy.playBlockBreakSound(xTile, yTile, zTile, block, this.dimension);
+                        	if(valuesArray[2] == 1) worldObj.setBlockToAir(xTile, yTile, zTile);  
```

Comment:
I think it's a good idea to check this against world protection like WG to make sure it doesn't bypass it - you've got a bukkit server iirc?

Reply:
Is it even possible to check for WorldGuard protection from within a forge mod? But it got me thinking, it's probably a good idea to add a 'allowBlockDestruction' field to the BulletHitEvent.
Maybe even a 'BlockPenetrationEvent' where you could cancel the destruction of a block?
What do you think?

Reply:
A block break event might be a good idea, yeah. I'l think about that.
In regards to world guard. I just tested it with my own regions plugin, and flans mod simply bypasses the bukkit event. I don't really know if there is a fix for that though unless we implement a BlockBreakEvent which savvy server owners could then cancel...

Reply:
Hm, that's a problem then. Does /teams vehiclesBreakBlocks bypass it too? I know explosions don't, because it uses the default explosion which bukkit patches its events into. I think it'd be a good idea to investigate how we could fire those events ourselves in the future, but for now I'm happy to accept it with a warning in the changelog, and block breaking disabled by default.

Reply:
Yep, vehiclesBreakBlocks also bypasses bukkit. And I agree, in the future we should look into this

Reply:
This is unrelated now, but I seem to remember bukkit inserts its own events into default minecraft code - when those get overriden by forge's patches, the events no longer fire. I suppose fixing those in cauldron might be the best way, but if we can sort something out in the mod (using normal bukkit events), that'd be awesome.